qualtran/bloqs/data_loading/qroam_clean.py

```diff
@@ -0,0 +1,389 @@
+                for sel_k in np.ndindex(block_sizes):
+                    sel_kwargs = {reg.name: sel for reg, sel in zip(swz.selection_registers, sel_k)}
+                    curr_data = swz.call_classically(**sel_kwargs, targets=np.copy(data[sel_l]))[-1]
+                    idx = (*sel_l, *sel_k)
+                    permuted_batched_data[idx][:] = curr_data
+
+            n_blocks = len(self.block_sizes)
+            transpose_axes = [x for i in range(n_blocks) for x in [i, i + n_blocks]]
+            transpose_axes += [i + 2 * n_blocks for i in range(n_blocks)]
+            desired_shape = tuple(
+                NbyK * K for NbyK, K in zip(self.batched_qrom_shape, self.block_sizes)
+            )
+            ret.append(
+                permuted_batched_data.transpose(transpose_axes).reshape(
+                    desired_shape + self.block_sizes
+                )
+            )
+        return ret
+
+    @cached_property
+    def junk_registers(self) -> Tuple[Register, ...]:
+        # The newly allocated registers should be kept around for measurement based uncomputation.
+        junk_regs = []
+        block_size = prod(self.block_sizes)
+        for reg in self.target_registers:
+            assert reg.shape == ()
+            if is_symbolic(block_size) or block_size > 1:
+                junk_regs += [attrs.evolve(reg, name='junk_' + reg.name, shape=(block_size - 1,))]
+        return tuple(junk_regs)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        ret: Dict[Bloq, SymbolicInt] = defaultdict(lambda: 0)
+        ret[self.qrom_bloq] += 1
+        for swz in self.swap_with_zero_bloqs:
+            if any(is_symbolic(s) or s > 0 for s in swz.selection_bitsizes):
+                ret[swz] += 1
+        return set(ret.items())
+
+    def _build_composite_bloq_with_swz_clean(
+        self,
+        bb: 'BloqBuilder',
+        ctrl: List['SoquetT'],
+        selection: List['SoquetT'],
+        qrom_targets: List['SoquetT'],
+    ) -> Tuple[List['SoquetT'], List['SoquetT'], List['SoquetT']]:
+        sel_l, sel_k = self._partition_sel_register(bb, selection)
+        ctrl, sel_l, qrom_targets = self._add_qrom_bloq(bb, ctrl, sel_l, qrom_targets)
+        sel_k, qrom_targets = self._add_swap_with_zero_bloq(bb, sel_k, qrom_targets)
+        selection = self._unpartition_sel_register(bb, sel_l, sel_k)
+        return ctrl, selection, qrom_targets
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> Dict[str, 'SoquetT']:
+        # Get the ctrl and target register for the SelectSwapQROM.
+        ctrl = [soqs.pop(reg.name) for reg in self.control_registers]
+        selection = [soqs.pop(reg.name) for reg in self.selection_registers]
+        if is_symbolic(*self.block_sizes):
+            raise ValueError(
+                f"Cannot decompose QROAM bloq with symbolic block sizes. Found {self.block_sizes=}"
```

Comment:
Is that `=` meant to be there?

Reply:
`f'{a=}'` in an f-string prints `'a=(value of a)'` whereas `f'{a}'` only prints `'(value of a)'`. The former is nicer IMO, so yes it's meant to be there.

Reply:
Neat. (I checked it locally but mistyped f'{a}=')